SYSTEM RESTORE REPLICA replica [ON CLUSTER cluster]

myrrc opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Provides a way to restore replicated table when the data is (possibly) present, but the ZooKeeper metadata is lost.
Resolves #13458.

Commit:
initial commit: add setting and stub

Commit:
typo

Commit:
Merge branch 'master' into feature/zk-system-recover-replica-option

Commit:
added test stub

amosbird commented:
I wish this feature was landed earlier lol.  I just went through some tedious time to move around terabytes of data via parts attaching yesterday...

Commit:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

Commit:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

Commit:
fix

Commit:
wip merging new integration test and code proto

filimonov commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -343,6 +343,48 @@ void InterpreterSystemQuery::restoreReplica(ASTSystemQuery & query)
+ // ALTER TABLE table_repl MOVE PARTITION 1 TO TABLE table_repl_rec;
+ // ALTER TABLE table_repl MOVE PARTITION 2 TO TABLE table_repl_rec;
+ //
+ // Alternatively, you can run in bash:clickhouse-client --format=TSVRaw -q"SELECT 'ALTER TABLE ' || database || '.' || table || ' MOVE PARTITION ID \'' || partition_id || '\' TO TABLE ' || table || '_rec'  || ';\n' FROM system.parts WHERE database = 'default' AND table = 'table_repl' AND active GROUP BY database, table, partition_id ORDER BY database, table, partition_id;" | clickhouse-client -m ...
```
It will create new part numbers. I think we need to register parts with existing numbers to avoid full resync.

Commit:
adding steps interpreters

Commit:
adding firstly proposed solution (moving parts etc)

Commit:
added checking zookeeper path existence

Commit:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

alexey-milovidov commented:
@excitoon What it happening here? Can we remove `do not test` label?

Commit:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

Commit:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

Commit:
fixing the include

Commit:
fixing and sorting includes

Commit:
fixing outdated struct

Commit:
fix the name

Commit:
added ast ptr as level of indirection

Commit:
fix ref

Commit:
updating the changes

Commit:
Merge branch 'feature/zk-system-recover-replica-option' of github:myrrc/clickhouse into feature/zk-system-recover-replica-option

Commit:
working on test stub

Commit:
fix iterator -> reference

Commit:
revert rocksdb submodule update

Commit:
fixed show privileges test

Commit:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

Commit:
updated the test stub

Commit:
replaced rand() with thread_local_rng(), updated the tests

updated the test

fixed test config path

test fix

removed error messages

fixed the test

updated the test

fixed string literal

fixed literal

typo: =

Commit:
fixed the empty replica error message

Commit:
updated the test and the code with logs

Commit:
updated the possible test cases, updated

Commit:
added the code/test milestone comments

Commit:
updated the test (added more testcases)

Commit:
replaced native assert with CH one

Commit:
individual replicas recursive delete fix

Commit:
updated the AS db.name AST

Commit:
two small logging fixes

Commit:
manually generated AST fixes

myrrc commented:
Depends on https://github.com/ClickHouse/ClickHouse/pull/18978.

myrrc committed:
Merge branch 'feature/attach-partition-local' into feature/zk-system-recover-replica-option

myrrc committed:
Updated the test, added the possible algo change

myrrc committed:
Some thoughts about optimizing the solution:

ALTER MOVE PARTITION .. TO TABLE -> move to detached/ + ALTER ... ATTACH

myrrc committed:
fix

myrrc committed:
Removed the replica sync in test as it's invalid

myrrc committed:
Some test tweaks

myrrc committed:
tmp

myrrc committed:
Rewrote the algo by using the executeQuery instead of

hand-crafting the ASTPtr.

Two questions still active.

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

myrrc committed:
tr: logging active parts

myrrc committed:
Extracted the parts moving algo into a separate helper function

myrrc committed:
Merge branch 'feature/attach-partition-local' into feature/zk-system-recover-replica-option

myrrc committed:
Merge branch 'feature/attach-partition-local' into feature/zk-system-recover-replica-option

myrrc committed:
Fixed the test data and the queries slightly

myrrc committed:
Replaced query to system.parts to direct invocation,

started building the test that breaks on various parts.

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

myrrc committed:
Added the case for tables when at least one replica is alive

myrrc committed:
Updated the test to test replicas restoration by detaching/attaching

myrrc committed:
Altered the test to check restoration without replica restart

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

myrrc committed:
Added the tables swap in the start if the server failed last time

myrrc committed:
Hotfix when only /replicas/replica... path was deleted

myrrc committed:
Restore ZK paths while creating a replicated MergeTree table

myrrc committed:
Updated the docs, fixed the algo for individual replicas restoration case

myrrc committed:
Initial parts table storage fix, tests sync fix

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

myrrc committed:
Reverted individual replica restoration to general algo

myrrc committed:
Slightly optimised getDataParts

myrrc committed:
Trying another solution with parts detaching

myrrc commented:
So, the current more or less working solution consists of attaching local data from the first replica and re-downloading all of its parts from other replicas.

I tried playing with ZK to push local parts data, no luck. The main issue here is that straight after table initialization another thread invokes cloneReplicaIfNeeded which downloads all parts from other replica. Local parts are marked "unexpected" (as they're not present in ZK) and ignored.

Will try another solution with inital replica parts DETACH (so as other replicas will initialize as empty tables) and then perform ATTACH.

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

myrrc committed:
Rewrote algo without any steps, added ON CLUSTER support

myrrc committed:
Attaching parts from other replica on restoration

myrrc committed:
Getting part checksums from ZK

myrrc committed:
Removed ON CLUSTER, finished working solution

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

myrrc committed:
Multiple small changes after review

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -423,18 +433,20 @@ StorageReplicatedMergeTree::StorageReplicatedMergeTree(
-    else
+    } else
     {
-        /// In old tables this node may missing or be empty
```
Is the comment outdated?

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -2776,7 +2787,37 @@ void StorageReplicatedMergeTree::cloneReplica(const String & source_replica, Coo
         LogEntry log_entry;
-        log_entry.type = LogEntry::GET_PART;
+
+        if (!are_restoring_replica)
+            log_entry.type = LogEntry::GET_PART;
+        else
```
```
if (condition)
    do_something();
else
    do_something_else();
```

is usually more readable than

```
if (!condition)
    do_something_else();
else
    do_something();
```

myrrc replied:
Sure, in your example. But I tend to place branches with less code in the top, so it's more like

```cpp
if (!condition)
  do_stuff();
else
{
  do_a_lot_of_stuff();
  do_something_else();
}
```

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -5028,6 +5071,46 @@ bool StorageReplicatedMergeTree::getFakePartCoveringAllPartsInPartition(const St
+
+        if (const String path = relative_data_path + "detached/" + part->getRelativePathForPrefix("");
+            disk->exists(path))
+            disk->removeRecursive(path);
```
Let's write warning in log

tavplubix commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -5028,6 +5071,46 @@ bool StorageReplicatedMergeTree::getFakePartCoveringAllPartsInPartition(const St
+        for (const auto& part : parts)
+            attachPartition(std::make_shared<ASTLiteral>(part->name), metadata_snapshot, true, getContext());
+
+    are_restoring_replica = true;
+
+    startup();
+
+    are_restoring_replica = false;
```
Btw, it may be a good idea to use node in zk instead of flag in memory

myrrc replied:
Nope, but RESTORE is not supposed to be invoked from multiple threads, I guess.

tavplubix replied:
`restoreMetadataOnReadonlyTable()` and `cloneReplica()` are called from different threads, aren't they? But protection from concurrent `RESTORE` queries is also useful (we can just throw exception if replica is restoring now).

myrrc replied:
Changed to use atomics.

myrrc commented on docs/en/sql-reference/statements/system.md:
```diff
@@ -269,12 +270,46 @@ After running this statement the `[db.]replicated_merge_tree_family_table_name`
+### RESTORE REPLICA {#query_language-system-restore-replica}
+
+Restores a replica if data is [possibly] present but Zookeeper metadata is lost.
+Works when ZooKeeper root `/` is lost. 
```
No, it does not, and I stated it explicitly. If table path `/replicas/replica_name` is lost, but there exists an alive replica, calling `RESTORE` would result in data duplication.

tavplubix replied:
By the words "the whole table path" I mean `zookeeper_path`, not `replica_path`. But even for lost `replica_path` it works in your test...

myrrc replied:
No, it does not.
For test https://github.com/ClickHouse/ClickHouse/pull/13652/commits/ce2f1600b57a16cd6d62984d4e081fc956a4ef67#diff-0e7bd00c923561fa5d8095062f16635e6104089909f7afd53e29d9e8dc12b2bfL53

It duplicated all parts on replicas 2 and 3 (well, it registered old present paths as new ones as ZK data was lost).

myrrc commented on programs/CMakeLists.txt:
```diff
@@ -33,6 +33,7 @@ option (ENABLE_CLICKHOUSE_OBFUSCATOR "Table data obfuscator (convert real data t
     ${ENABLE_CLICKHOUSE_ALL})
 
 # https://clickhouse.tech/docs/en/operations/utilities/odbc-bridge/
+# TODO Also needs NANODBC.
```
If you build ClickHouse with `-DENABLE_LIBRARIES=OFF -DENABLE_ODBC=ON`, you won't be able to build the ODBC bridge as it also depends on `-DENABLE_NANODBC` whic is not checked here.

tavplubix replied:
Nice catch! (I thought this line was changed accidentally, because it looks unrelated)
cc: @kssenii

myrrc commented on src/Storages/MergeTree/MergeTreeData.h:
```diff
@@ -401,7 +401,24 @@ class MergeTreeData : public IStorage, public WithContext
 
     /// Returns a copy of the list so that the caller shouldn't worry about locks.
-    DataParts getDataParts(const DataPartStates & affordable_states) const;
+    template <class ...AffordableStates>
+    inline DataParts getDataParts(AffordableStates ...states) const
```
A small speedup due to no dynamic memory allocation in vector.

tavplubix replied:
Do you have some performance tests?
I don't like templates, especially variadic templates, especially unwarranted variadic templates :)
And I don't think that we should make code less readable for 0.00000001% of performance.

myrrc replied:
Well, It makes code more readable as for me, but if you stand your ground about that, I'll change.

myrrc commented on tests/integration/test_restore_replica/test.py:
```diff
@@ -0,0 +1,91 @@
+
+    assert_eq_with_retry(node_2, "SELECT sum(n), count() FROM test", res)
+    assert_eq_with_retry(node_3, "SELECT sum(n), count() FROM test", res)
+
```
This query does not work with ON CLUSTER as ZK metadata is lost, so I excluded both the code and the tests.

myrrc replied:
You can checkout at https://github.com/ClickHouse/ClickHouse/pull/13652/commits/ce2f1600b57a16cd6d62984d4e081fc956a4ef67 and see -- ON CLUSTER with DDL guards yielded an eternal timeout.

tavplubix replied:
Could you please explain in more details why? `zk.delete("/clickhouse/tables/test", recursive=True)` should not affect DDL queue of `ON CLUSTER` queries, it should work.

tavplubix replied:
Hm, it's interesting, I will take a look

myrrc commented on src/Interpreters/InterpreterSystemQuery.cpp:
```diff
@@ -423,12 +426,52 @@ BlockIO InterpreterSystemQuery::execute()
+    auto * const table_replicated_ptr = dynamic_cast<StorageReplicatedMergeTree *>(table_ptr.get());
+
+    if (table_replicated_ptr == nullptr)
+        throw Exception(ErrorCodes::UNKNOWN_TABLE, "There is no replicated table \"{}.{}\"", db_name, table_name);
```
Current version is consistent with https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/InterpreterSystemQuery.cpp#L395

tavplubix replied:
Ok, let's change both

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

myrrc committed:
Fixing parallel test

myrrc committed:
Supporting rewritten form on cluster

myrrc committed:
Test fix

myrrc committed:
Moar logging

myrrc committed:
Using source replica as checksum provider

tavplubix committed:
improve test, remove some code from parser

myrrc committed:
Trying solution with move to detached + forget

myrrc committed:
Moving all parts (not only Committed) to detached

myrrc committed:
Merge remote-tracking branch 'upstream/master' into feature/zk-system-recover-replica-option

myrrc committed:
Edited docs for RESTORE REPLICA

myrrc committed:
Re-merging

tavplubix committed:
minor fixes

Slach commented:
@myrrc ;) kudos and congratulation!

leiless commented:
Hi all, which ClickHouse version included this `SYSTEM RESTORE REPLICA replica [ON CLUSTER cluster]` feature?

Slach commented:
@leiless 21.7+ look tags on https://github.com/ClickHouse/ClickHouse/commit/4c391f8e9945df67b171d20612b62b45451ea8b4

leiless commented:
Hi all again, what are the typical use scenarios of `SYSTEM RESTORE REPLICA replica [ON CLUSTER cluster]`?
Is it when we have a damaged/out-of-the sync zookeeper, ClickHouse may start in `readonly` state, which we utilize `SYSTEM RESTORE REPLICA ...` for remedy?

blafry commented:
@Slach Is there a chance that this feature will be backported to LTS version 21.3?

tavplubix commented:
> Is there a chance that this feature will be backported to LTS version 21.3?

No, we backport bugfixes only, features are never backported
